### ISSUE TITLE
CD4021 Input Shift Register doesn't like optimization

### ISSUE THREAD

**Issue by stephenhensley** [electro-smith/libDaisy#237]
I'm guessing its optimizing out one of the loops or moving some of the GPIO stuff around. Because pushing one button reads in all of them. Setting to `-O1` or `-O0` both work, but `-O2` and above do not.

Should be able to put a dsb or something there to make sure it does everything before moving on.

**Comment by stephenhensley** [comment]
Just a note in case my changes get lost in the mess of PRs I'm about to merge before addressing this directly, but there was no delay between the clock strobe on the back end of the for loop for each data bit. This was either getting optimized out, or it was just going high for so little time that it `wasn't long for the 4021 to catch.

Here's the functional update function:

```C++
// 3000 = 15us max clock/rise/fall time
static const uint32_t kGpioDelayAmt = 1; // in ticks @ 200MHz
void dsy_sr_4021_update(dsy_sr_4021_handle *sr)
{
    uint8_t idx;
    // Strobe CS Pin
    dsy_gpio_write(&sr->clk, 0);
    dsy_gpio_write(&sr->cs, 1);
    dsy_tim_delay_tick(kGpioDelayAmt);
    dsy_gpio_write(&sr->cs, 0);
    // Clock through data.
    for(uint8_t i = 0; i < 8 * sr->num_daisychained; i++)
    {
        // Grab data from each parallel data pin.
        dsy_gpio_write(&sr->clk, 0);
        dsy_tim_delay_tick(kGpioDelayAmt);
        for(uint8_t j = 0; j < sr->num_parallel; j++)
        {
            idx = (8 * sr->num_daisychained - 1) - i;
            idx += (8 * sr->num_daisychained * j);
            sr->states[idx] = dsy_gpio_read(&sr->data[j]);
        }
        dsy_gpio_write(&sr->clk, 1);
        dsy_tim_delay_tick(kGpioDelayAmt);
    }
}
```

**Comment by stephenhensley** [comment]
this was resolved in #244
